website/blog/2025-08-30-mlflow-prompt-evaluate/index.md

```diff
@@ -0,0 +1,503 @@
+
+### 1. Set Up
+
+Install the required packages:
+
+```bash
+pip install openai mlflow tiktoken aiohttp -qU
+```
+
+**openai:** For interacting with OpenAI models and APIs  
+**mlflow:** For experiment tracking, model management, and GenAI workflow tools  
+**tiktoken:** For efficient tokenization, especially useful with OpenAI models  
+**aiohttp:** For asynchronous HTTP requests, enabling efficient API calls
+
+For this tutorial we use OpenAI, but the approach extends to other LLM providers.
+We can prompt the user to type the OpenAI API key without echoing using getpass():
+
+```python
+import os
+from getpass import getpass
+
+os.environ["OPENAI_API_KEY"] = getpass("Your OpenAI API Key: ")
+```
+
+### 2. Observe the Data
+
+Let's read a randomly selected annotated file. The following utils functions faciliate this task.
+
+```python
+import json
+import random
+from PIL import Image as im
+import base64
+from io import BytesIO
+import pandas as pd
+import re
+from typing import Any
+
+DATA_DIRECTORY = "./data" # for simplicity, a local directory
+ANNOTATIONS_DIRECTORY = os.path.join(DATA_DIRECTORY, "annotations")
+IMAGES_DIRECTORY = os.path.join(DATA_DIRECTORY, "images")
+
+def _extract_qa_pairs(data: dict) -> list[tuple[str, str]]:
+    """Extracts question-answer pairs from OCR-style linked data as a list of tuples"""
+    qa_pairs = []
+    elements = {item["id"]: item for item in data}
+
+    for item in data:
+        if item["label"] != "question" or not item["linking"]:
+            continue
+        for q_id, a_id in item["linking"]:
+            if q_id != item["id"]:
+                continue
+            answer = elements.get(a_id)
+            if answer and answer["label"] == "answer":
+                q_text = " ".join(w["text"] for w in item["words"])
+                a_text = " ".join(w["text"] for w in answer["words"])
+                qa_pairs.append((q_text, a_text))
+
+    return qa_pairs
+
+def get_json(file_name: str, directory: str = ANNOTATIONS_DIRECTORY) -> dict:
+    file_name += ".json" if not file_name.endswith(".json") else file_name
+    path = os.path.join(directory, file_name)
+    with open(path, "r", encoding="utf-8") as f:
+        contents = json.load(f)["form"]
+        if isinstance(contents, list):
+            if all(isinstance(page, list) for page in contents):
+                flat_items = [item for page in contents for item in page]
+            else:
+                flat_items = contents
+            tuples_result = _extract_qa_pairs_tuple(flat_items)
+        else:
+            tuples_result = _extract_qa_pairs_tuple(contents)
+
+        # Convert list of tuples to dict before returning
+        return {key.rstrip(':'): value for key, value in tuples_result}
+
+def get_random_files(
+    directory: str = ANNOTATIONS_DIRECTORY, n: int = 1
+) -> list[str]:
+    if not os.path.exists(directory):
```

Comment:
Can probably remove this. The exception thrown from `os.listdir(<non-existent-directory>)` is pretty clear and easy to troubleshoot.